### ISSUE TITLE
Running 'bin/magento catalog:images:resize' doesn't check if an image was already resized before, therefore it is superslow as it always resizes all images

### ISSUE THREAD

**Issue by hostep** [magento/magento2#26796]
I'm pretty sure in Magento 2.1 (and maybe 2.2) when you ran `bin/magento catalog:images:resize`, it only resized images which weren't already resized before.

This no longer seems to be the case in Magento 2.3.
I think this behavior got introduced by https://github.com/magento/magento2/commit/498879059b400f8785ee63ba371798ffec33ae53 (not verified yet)

### Preconditions (*)
1. Magento 2.3.3
2. PHP 7.2.26

### Steps to reproduce (*)
1. Have a shop with a product with an image
2. Run `bin/magento catalog:images:resize`
3. Look at the timestamp of the resized images in `pub/media/catalog/product/cache/`
4. Wait a while
5. Run `bin/magento catalog:images:resize` again
6. Notice that the modified timestamp of images in `pub/media/catalog/product/cache/` has been altered, which means the images got resized again even if they already exist.

### Expected result (*)
1. Executing `bin/magento catalog:images:resize` should not re-save an already resized image

### Actual result (*)
1. Executing `bin/magento catalog:images:resize` re-saves already resized images

### Temp solution
This seems to resolve the issue after some very initial testing:
```diff
diff --git a/app/code/Magento/MediaStorage/Service/ImageResize.php b/app/code/Magento/MediaStorage/Service/ImageResize.php
index 6da48ee69ed..ccfce994752 100644
--- a/app/code/Magento/MediaStorage/Service/ImageResize.php
+++ b/app/code/Magento/MediaStorage/Service/ImageResize.php
@@ -287,13 +287,18 @@ class ImageResize
     private function resize(array $viewImage, string $originalImagePath, string $originalImageName)
     {
         $imageParams = $this->paramsBuilder->build($viewImage);
-        $image = $this->makeImage($originalImagePath, $imageParams);
         $imageAsset = $this->assertImageFactory->create(
             [
                 'miscParams' => $imageParams,
                 'filePath' => $originalImageName,
             ]
         );
+        if ($this->mediaDirectory->isFile($imageAsset->getPath())) {
+            // image was already resized, let's quit!
+            return;
+        }
+
+        $image = $this->makeImage($originalImagePath, $imageParams);

         if (isset($imageParams['watermark_file'])) {
             if ($imageParams['watermark_height'] !== null) {
```

Postponing the initialization of the `$image` variable also has a very big impact on performance.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @shikhamis11 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-31322` were created

**Issue Available**: @shikhamis11, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by hostep** [comment]
Thanks @shikhamis11 for verifying!

I just tested some more on different clean Magento installations:
- 2.1.18: images which were resized before aren't being saved again => **good**
- 2.2.11: images which were resized before are being saved again => **not good**
- 2.3.4: images which were resized before are being saved again => **not good**

**Comment by gwharton** [comment]
Please make sure any mods to this area also work in database media storage mode. It is often overlooked.

**Comment by hostep** [comment]
@gwharton: I was aware of your recent changes in that area regarding the database storage, so I tried to take that into account in https://github.com/magento/magento2/pull/26801, but haven't actually tested it for real yet.

**Comment by gwharton** [comment]
Ahh yes. I should have spotted that in the PR before commenting. Changes look good, although also untested here.

**Comment by engcom-Alfa** [comment]
The issue is still reproducible on 2.4-develop

**Comment by swell-d** [comment]
@hostep Many thanks! Great job

**Comment by hostep** [comment]
https://github.com/magento/magento2/pull/26801 got merged a while ago, so we can close this issue.

No information was provided about in which Magento versions it will be fixed, but I'm guessing 2.4.0 and maybe 2.3.6?
